lib/client/doc.js

```diff
@@ -413,30 +410,66 @@ Doc.prototype.fetch = function(callback) {
 
 // Fetch the initial document and keep receiving updates
 Doc.prototype.subscribe = function(callback) {
-  this.wantSubscribe = true;
-  if (this.connection.canSend) {
-    var isDuplicate = this.connection.sendSubscribe(this);
-    pushActionCallback(this.inflightSubscribe, isDuplicate, callback);
-    return;
-  }
-  this.pendingFetch.push(callback);
+  var wantSubscribe = true;
+  this._queueSubscribe(wantSubscribe, callback);
 };
 
 // Unsubscribe. The data will stay around in local memory, but we'll stop
 // receiving updates
 Doc.prototype.unsubscribe = function(callback) {
-  this.wantSubscribe = false;
-  // The subscribed state should be conservative in indicating when we are
-  // subscribed on the server. We'll actually be unsubscribed some time
-  // between sending the message and hearing back, but we cannot know exactly
-  // when. Thus, immediately mark us as not subscribed
-  this.subscribed = false;
+  var wantSubscribe = false;
+  this._queueSubscribe(wantSubscribe, callback);
+};
+
+Doc.prototype._queueSubscribe = function(wantSubscribe, callback) {
+  var nextRequest = this.inflightSubscribe || this.pendingSubscribe[0];
+  var hasOtherRequests = this.pendingSubscribe.some(function(request) {
+    return request !== nextRequest;
+  });
+  if (nextRequest && nextRequest.wantSubscribe === wantSubscribe && !hasOtherRequests) {
```

Comment:
This will only coalesce if there's a single next request. As discussed, it should be OK to coalesce in the case of `s -> us -> s -> s`, by using the latest request instead of the next request. Then we don't need to check for other requests, either.